### PR TITLE
Aurora native ERC-20 connector

### PR DESCRIPTION
**Pull request by karim-en**
Implement token connector for native erc20 aurora tokens using asyncfy approach.
Implemented:
- Native erc20 token factory: near contract
- Native erc20 locker: eth contract
- Async aurora: near contract

### CODE REVIEW

File: async-aurora/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+codegen-units = 1
+# Tell `rustc` to optimize for small code size.
+opt-level = "z"
+lto = true
+debug = false
+panic = "abort"
+overflow-checks = true
+
+[dependencies]
+near-sdk = { git = "https://github.com/near/near-sdk-rs", rev = "613adcab6322e4f830efb4d5bde2a810ae642158" }
```
Comment (sept-en) [inline async-aurora/Cargo.toml:20]: The commit matches some random commit. Should be pointer to stable release.

```suggestion
near-sdk = { git = "https://github.com/near/near-sdk-rs", tag = "3.1.0" }
```

File: aurora-engine-parameters/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+[package]
+name = "aurora-engine-parameters"
+version = "0.1.0"
+edition = "2021"
+
+[dependencies]
+near-sdk = { git = "https://github.com/near/near-sdk-rs", rev = "613adcab6322e4f830efb4d5bde2a810ae642158" }
```
Comment (sept-en) [inline aurora-engine-parameters/Cargo.toml:7]: Same here.

```suggestion
near-sdk = { git = "https://github.com/near/near-sdk-rs", tag = "3.1.0" }
```

File: bridge-aurora-token-factory/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+uint = { version = "0.8.3", default-features = false }
+rlp = { version = "0.5.0", default-features = false }
+hex = "0.4.2"
+getrandom = { version = "0.2.6", features = ["js"] }
+primitive-types = { version = "0.10.0", default-features = false, features = ["rlp"] }
+aurora-engine-parameters = { path = "../aurora-engine-parameters", default-features = false }
+sha3 = { version = "0.9.1", default-features = false }
+
+[dev-dependencies]
+near-sdk-sim = { git = "https://github.com/near/near-sdk-rs", rev = "613adcab6322e4f830efb4d5bde2a810ae642158" }
```
Comment (sept-en) [inline bridge-aurora-token-factory/Cargo.toml:38]: And here

```suggestion
near-sdk-sim = { git = "https://github.com/near/near-sdk-rs", tag = "3.1.0" }
```

File: bridge-aurora-token-factory/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ethabi = { git = "https://github.com/darwinia-network/ethabi", branch = "xavier-no-std", default-features = false }
+serde = { version = "*", features = ["derive"] }
+serde_json = "*"
+uint = { version = "0.8.3", default-features = false }
+rlp = { version = "0.5.0", default-features = false }
+hex = "0.4.2"
+getrandom = { version = "0.2.6", features = ["js"] }
+primitive-types = { version = "0.10.0", default-features = false, features = ["rlp"] }
+aurora-engine-parameters = { path = "../aurora-engine-parameters", default-features = false }
+sha3 = { version = "0.9.1", default-features = false }
```
Comment (sept-en) [inline bridge-aurora-token-factory/Cargo.toml:35]: This shouild be moved to dev-dependencies

**Review by sept-en — CHANGES_REQUESTED** [review]

**Review by mfornet — CHANGES_REQUESTED** [review]
Scope for `async-aurora` goes beyond Aurora Native ERC20 connector, which IMO is a good reason to:
- Implement it in a separate repository (actually this is more related to the engine than to the rainbow-bridge)
- Provides specific documentation and tools about how it can be integrated with other third party tools.

File: bridge-aurora-token-factory/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,39 @@
+[package]
+name = "bridge-aurora-token-factory"
+version = "0.1.5"
+authors = ["Near Inc <hello@near.org>"]
+edition = "2018"
+publish = false
+
```
Comment (sept-en) [inline bridge-aurora-token-factory/Cargo.toml:7]: Please add a description

File: bridge-aurora-token-factory/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,39 @@
+[package]
+name = "bridge-aurora-token-factory"
+version = "0.1.5"
+authors = ["Near Inc <hello@near.org>"]
```
Comment (sept-en) [inline bridge-aurora-token-factory/Cargo.toml:4]: The author should be Aurora

File: bridge-aurora-token-factory/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+edition = "2018"
```
Comment (sept-en) [inline bridge-aurora-token-factory/Cargo.toml:5]: why not 2021?

File: erc20-connector/contracts/AuroraERC20Locker.sol [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/utils/math/SafeMath.sol";
+import "@openzeppelin/contracts/utils/Strings.sol";
+import "rainbow-bridge-sol/nearbridge/contracts/AdminControlled.sol";
+import "./AccountIds.sol";
+
+contract AuroraERC20Locker is AdminControlled, AccountIds {
```
Comment (sept-en) [inline erc20-connector/contracts/AuroraERC20Locker.sol:12]: Need to add UUPS + ACL here. Could be in a separate follow-up PR though.

Comment (sept-en) [inline erc20-connector/contracts/AuroraERC20Locker.sol:12]: Is there any objection to having this contract inside `erc20-connector` folder with others? I don't think so and IMO it's better to keep it in a separate directory, especially as there are no contract intersections.

**Review by sept-en — CHANGES_REQUESTED** [review]
Added some more comments and questions for discussions. Also, please add README for aurora native token connector with appropriate references to cross-contract-call and native token connector design discussions.

File: erc20-connector/contracts/AuroraERC20Locker.sol (hunk 8 above)
Comment (karim-en) [reply erc20-connector/contracts/AuroraERC20Locker.sol]: Fixed at 2271dc717d43bcbd80d1ea824a15211039fbd2e8

File: erc20-connector/contracts/AuroraERC20Locker.sol [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+    {
+        require(
+            compareStrings(predecessorAccountId(), nearTokenFactory),
+            "Mismatch factory account"
+        );
+        LockEvent memory lockEvent = lockEvents[nonce];
+        require(lockEvent.token == token, "Mismatch token address");
+        amount = lockEvent.amount;
+        receipient = lockEvent.receipient;
+        delete lockEvents[nonce];
```
[the remark this answers is not in the document]
Comment (karim-en) [reply erc20-connector/contracts/AuroraERC20Locker.sol]: Added at f005b19c3e73964e102018e69c462aaeffa9d7eb

File: bridge-aurora-token-factory/Cargo.toml (hunk 7 above)
Comment (karim-en) [reply bridge-aurora-token-factory/Cargo.toml]: Fixed in f005b19c3e73964e102018e69c462aaeffa9d7eb

File: bridge-aurora-token-factory/Cargo.toml (hunk 6 above)
Comment (karim-en) [reply bridge-aurora-token-factory/Cargo.toml]: Fixed in f005b19c3e73964e102018e69c462aaeffa9d7eb

File: bridge-aurora-token-factory/Cargo.toml (hunk 5 above)
Comment (karim-en) [reply bridge-aurora-token-factory/Cargo.toml]: Fixed in f005b19c3e73964e102018e69c462aaeffa9d7eb

File: erc20-connector/contracts/AuroraERC20Locker.sol [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+    ) public AdminControlled(_admin, _pausedFlags) {
+        lastLockEventIndex = 0;
+        nearTokenFactory = _nearTokenFactory;
+    }
+
+    function lockToken(
+        address ethToken,
+        uint256 amount,
+        string memory recipient
+    ) public pausable(PAUSED_LOCK) returns (string memory) {
```
[the remark this answers is not in the document]
Comment (karim-en) [reply erc20-connector/contracts/AuroraERC20Locker.sol]: Added in f005b19c3e73964e102018e69c462aaeffa9d7eb

File: erc20-connector/contracts/AuroraERC20Locker.sol [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+            string(
+                abi.encodePacked(
+                    "promises:",
+                    nearTokenFactory,
+                    "#",
+                    "deposit",
+                    "#",
+                    args,
+                    "#",
+                    "20000000000000"
```
[the remark this answers is not in the document]
Comment (karim-en) [reply erc20-connector/contracts/AuroraERC20Locker.sol]: Fixed in f005b19c3e73964e102018e69c462aaeffa9d7eb

File: erc20-connector/contracts/AuroraERC20Locker.sol [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+            msg.sender,
+            amount,
+            recipient
+        );
+        lockEvents[lastLockEventIndex] = locekEvent;
+        string memory args = string(
+            abi.encodePacked(
+                '{"token":"',
+                Strings.toHexString(uint160(ethToken), 20),
+                '","nonce":"',
```
[the remark this answers is not in the document]
Comment (karim-en) [reply erc20-connector/contracts/AuroraERC20Locker.sol]: Fixed in f005b19c3e73964e102018e69c462aaeffa9d7eb

File: erc20-connector/contracts/AuroraERC20Locker.sol [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+                Strings.toString(lastLockEventIndex),
+                '"}'
+            )
+        );
+
+        emit Locked(address(ethToken), msg.sender, amount, recipient);
```
[the remark this answers is not in the document]
Comment (karim-en) [reply erc20-connector/contracts/AuroraERC20Locker.sol]: Fixed in f005b19c3e73964e102018e69c462aaeffa9d7eb

File: erc20-connector/contracts/AuroraERC20Locker.sol [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+        IERC20(ethToken).safeTransferFrom(msg.sender, address(this), amount);
+        lastLockEventIndex++;
+        LockEvent memory locekEvent = LockEvent(
+            lastLockEventIndex,
+            ethToken,
+            msg.sender,
+            amount,
+            recipient
+        );
+        lockEvents[lastLockEventIndex] = locekEvent;
```
[the remark this answers is not in the document]
Comment (karim-en) [reply erc20-connector/contracts/AuroraERC20Locker.sol]: Fixed in f005b19c3e73964e102018e69c462aaeffa9d7eb

File: erc20-connector/contracts/AuroraERC20Locker.sol [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+
+    function claim(address token, uint256 nonce)
+        public
+        returns (uint256 amount, string memory receipient)
+    {
+        require(
+            compareStrings(predecessorAccountId(), nearTokenFactory),
+            "Mismatch factory account"
+        );
+        LockEvent memory lockEvent = lockEvents[nonce];
```
[the remark this answers is not in the document]
Comment (karim-en) [reply erc20-connector/contracts/AuroraERC20Locker.sol]: Changed in fe94840ff21bcb0a4196075317bae03c55e8fde8

File: bridge-aurora-token-factory/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,838 @@
...
+            contract: self.locker_address,
+            value: WeiU256::default(),
+            input: [LOCKER_CLAIM_SELECTOR, tail.as_slice()].concat(),
+        });    
+
+        ext_aurora::call(
+            call_args,
+            &self.aurora_account,
+            NO_DEPOSIT,
+            CLAIM_GAS,
```
[the remark this answers is not in the document]
Comment (karim-en) [reply bridge-aurora-token-factory/src/lib.rs]: Fixed in 6ed82190ae6e7a9a156108ba9777ef31b60395ec

File: erc20-connector/contracts/AuroraERC20Locker.sol (hunk 9 above)
[the remark this answers is not in the document]
Comment (karim-en) [reply erc20-connector/contracts/AuroraERC20Locker.sol]: I prefer to free app the unused storage to get a refund. 
https://soliditydeveloper.com/design-pattern-solidity-free-up-unused-storage

File: native-erc20-connector/contracts/AuroraERC20Locker.sol [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+    function claim(address token, uint256 lockEventIndex) public
+    {
+        require(
+            compareStrings(predecessorAccountId(), nearTokenFactory),
+            "Mismatch factory account"
+        );
+
+        LockEvent storage lockEvent = lockEvents[lockEventIndex];
+        require(lockEvent.index > 0, "The lockEventIndex is not exist");
+        require(lockEvent.token == token, "Mismatch token address");
```
Comment (sept-en) [inline native-erc20-connector/contracts/AuroraERC20Locker.sol:160]:
```suggestion
        require(lockEvent.token == token, "Token address doesn't match in the locked event");
```

File: bridge-aurora-token-factory/Cargo.toml [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+near-sdk = { git = "https://github.com/near/near-sdk-rs", rev = "613adcab6322e4f830efb4d5bde2a810ae642158" }
+near-contract-standards = { git = "https://github.com/near/near-sdk-rs", rev = "613adcab6322e4f830efb4d5bde2a810ae642158" }
+eth-types =  { git = "https://github.com/near/rainbow-bridge", rev = "9ffb08e6613788096f26636f695b4c7e5fa5b261" }
+admin-controlled = { git = "https://github.com/near/rainbow-bridge", rev = "9ffb08e6613788096f26636f695b4c7e5fa5b261" }
+tiny-keccak = "1.4.0"
+ethabi = "12"
+serde = { version = "*", features = ["derive"] }
+serde_json = "*"
+uint = { version = "0.8.3", default-features = false }
+rlp = "0.4.2"
```
Comment (sept-en) [inline bridge-aurora-token-factory/Cargo.toml:31]: I don't think this is used anywhere. Please check other dependencies too

File: sim-tests/Cargo.toml [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+publish = false
+autobenches = false
+
+[dependencies]
+aurora-engine-parameters = { path = "../aurora-engine-parameters", default-features = false }
+aurora-engine-types = { git = "https://github.com/aurora-is-near/aurora-engine", default-features = false, features = ["std"] }
+aurora-engine-sdk = { git = "https://github.com/aurora-is-near/aurora-engine", default-features = false, features = ["std"] }
+borsh = { version = "0.8.2", default-features = false }
+sha3 = { version = "0.9.1", default-features = false }
+rlp = { version = "0.5.0", default-features = false }
```
Comment (sept-en) [inline sim-tests/Cargo.toml:21]: I don't think this is used anywhere. Please check other dependencies too

**Review by sept-en — CHANGES_REQUESTED** [review]

File: sim-tests/tests/token_transfer.rs [hunk 20]
```diff
@@ -0,0 +1,507 @@
...
+            0,
+        )
+        .unwrap_json_value()
+        .as_str()
+        .unwrap()
+        .parse()
+        .unwrap()
+}
+
+#[test]
```
Comment (sept-en) [inline sim-tests/tests/token_transfer.rs:409]: Please add all these tests to CI.

**Comment by sept-en** [comment]
Closed as a stalled PR in favor of https://github.com/aurora-is-near/native-erc20-connector

**Pull request closed without merging**